frontend/src/lib/ReportNode.ts

```diff
@@ -0,0 +1,509 @@
+ * each of which corresponds to a tree mutation ("add an element",
+ * "add a block", "add rows to an existing element"). The frontend builds the
+ * tree bit by bit in response to these `Delta`s.
+ *
+ * To render the app, the `ReportView` class walks this tree and outputs
+ * a corresponding DOM structure, using React, that's essentially a mapping
+ * of `ReportElement` -> `ReactNode`. This rendering happens "live" - that is,
+ * the app is re-rendered each time a new `Delta` is received.
+ *
+ * Because the app gets re-rendered frequently, updates need to be fast.
+ * Our React components - the building blocks of the app - are "pure"
+ * (see https://reactjs.org/docs/react-api.html#reactpurecomponent), which
+ * means that React uses shallow comparison to determine which ReactNodes to
+ * update.
+ *
+ * Thus, each node in our tree is _immutable_ - any change to a `ReportNode`
+ * actually results in a *new* `ReportNode` instance. This occurs recursively,
+ * so inserting a new `ElementNode` into the tree will also result in new
+ * `BlockNode`s for each of that Element's ancestors, all the way up to the
+ * root node. Then, when React re-renders the app, it will re-traverse the new
+ * nodes that have been created, and rebuild just the bits of the app that
+ * have changed.
+ */
+export interface ReportNode {
+  /**
+   * The ID of the report this node was generated in. When a report finishes
+   * running, the app prunes all stale nodes.
+   */
+  readonly reportId: string
+
+  /**
+   * Return the ReportNode for the given index path, or undefined if the path
+   * is invalid.
+   */
+  getIn(path: number[]): ReportNode | undefined
+
+  /**
+   * Return a copy of this node with a new element set at the given index
+   * path. Throws an error if the path is invalid.
+   */
+  setIn(path: number[], node: ReportNode, reportId: string): ReportNode
+
+  /**
+   * Recursively remove children nodes whose reportID is no longer current.
+   * If this node should no longer exist, return undefined.
+   */
+  clearStaleNodes(currentReportId: string): ReportNode | undefined
+
+  /**
+   * Recursively add all Elements contained in the tree to a set.
+   */
+  getElements(elementSet: Set<Element>): void
+}
+
+/**
+ * A leaf ReportNode. Contains a single element to render.
+ */
+export class ElementNode implements ReportNode {
+  public readonly element: Element
+
+  public readonly metadata: ForwardMsgMetadata
+
+  public readonly reportId: string
+
+  /**
+   * A lazily-created immutableJS version of our element.
+   *
+   * This is a temporary! `immutableElement` is currently needed for
+   * dataframe-consuming elements because our dataframe API is
+   * immutableJS-based. It'll go away when we've converted to an ArrowJS-based
+   * dataframe API.
+   *
+   * Because most elements do *not* use the Dataframe API, and therefore
+   * do not need to access `immutableElement`, it is calculated lazily.
```

Comment:
It's to prevent foot-shooting, yeah. If you have a big dataframe, then toImmutableElement would be expensive, and it would be sort of an invisible landmine to do things like `this.node.immutableElement.get("some boolean") ? this.node.immutableElement.get("some value") : this.node.immutableElement.get("some other value")` or whatever. 

In practice I'm not sure if we do anything like this or not, but the complexity for caching it is very low.